### ISSUE TITLE
Bad eval results but good inference

### ISSUE THREAD

**Issue by SiBensberg** [google/automl#804]
Firstly thanks for your great work. 
The network actually predicts my custom dataset well. On inferenced data the detections are looking very good. Every labeled object is detected correctly. Now here my problem. I already mentioned it in another post but I think it is a problem for an own issue. The eval results are very bad ~0.06 and do not represent the good detections of the trained model (I think). I trained another architecture and the detections are looking worse but the eval shows AP50 of 0.85. 

| AP             |  Losses |
:-------------------------:|:-------------------------:
![image](https://user-images.githubusercontent.com/55738346/94137521-cf85e780-fe66-11ea-9be0-336da0ee2287.png)  |  ![image](https://user-images.githubusercontent.com/55738346/94137653-078d2a80-fe67-11ea-9be2-15f32e69e49b.png) |

I trained 2 networks one with standard parameters and one with tuned anchors both AP50 are at about 0.06 after a few epochs.

train command:
```
python3 main.py --mode=train_and_eval \
	    --training_file_pattern=/home/work/train*.record \  #just one file
	    --validation_file_pattern=/home/work/val*.record \  #just one file
	    --model_name=efficientdet-d1 \
	    --model_dir=efficientdet-d1-anchor \
	    --ckpt=efficientdet-d1 \
	    --train_batch_size=22 \
	    --eval_batch_size=22 --eval_samples=990 \
	    --hparams=voc_config.yaml \
	    --num_examples_per_epoch=22000 --num_epochs=300 \
	    --save_checkpoints_steps=1000
```

config.yaml:
```
num_classes: 5
var_freeze_expr: '(efficientnet)'
label_map: {1: one, 2: two, 3: three, 4: four}
use_keras_model: False 
anchor_scale: 1  # smallest object is about 8px
num_scales: 5 
aspect_ratios: [1.0, 0.4, 2.49]
optimizer: sgd
learning_rate: 0.03
lr_warmup_init: 0.003
jitter_min: 0.9 # the color is important for the detection of objects tuned this down
jitter_max: 1.1
max_instances_per_image: 200  # got an error here with 100

```

It seems the AP gets calculated correctly for others. So is the problem in my dataset? But why are the detections good then?

**Comment by mingxingtan** [comment]
It is hard to tell. Maybe you should first visualize your val data to ensure the groundtruth labels and image ids are correct?

**Comment by SiBensberg** [comment]
I took a look at the .records with this code [https://github.com/sulc/tfrecord-viewer]( https://github.com/sulc/tfrecord-viewer). It was easy and worked with only small changes. 
The val record data looks good. The gt labels are displayed correctly. I am not sure what you mean with the image id.

**Comment by SiBensberg** [comment]
I made the somewhat strange observation highering the eval_samples lowers the results. I mean of course the results are not precise anymore with that few samples. But the continual degression of precision makes me think there may be something wrong. There are 4000 samples in the .record file

| num_samples | AP   | AP50 |
|-------------|------|------|
| 25          | 0.63 | 0.96 |
| 50          | 0.50 | 0.72 |
| 100         | 0.34 | 0.51 |
| 200         | 0.27 | 0.39 |
| 300         | 0.25 | 0.37 |
| 1000        | 0.04 | 0.05 |

**Comment by SiBensberg** [comment]
I found the Problem. Do not use the fix for the `source_id` suggested in #307.
I just made a simple loop and gave every image in the record a new source ID. The Evaluation is now working.
now with all 4000 samples the AP50 is **0.806**

**Comment by kartik4949** [comment]
> I took a look at the .records with this code https://github.com/sulc/tfrecord-viewer. It was easy and worked with only small changes.
> The val record data looks good. The gt labels are displayed correctly. Sorry, I do not know what you mean with image id.
> Maybe it is important to mention that all my pictures have different resolutions.

maybe you can use this from now #865

**Comment by SiBensberg** [comment]
Thank you. This is actually a nice thing to have!

**Comment by JJangD** [comment]
> I found the Problem. Do not use the fix for the `source_id` suggested in #307.
> I just made a simple loop and gave every image in the record a new source ID. The Evaluation is now working.
> now with all 4000 samples the AP50 is **0.806**

You mean you gave every image an all different new source ID? ex) 1,2,3,4, ...
Not as #307 which made all source to ID zero.

**Comment by SiBensberg** [comment]
@JJangD Exactly. If I understood right, while evaling all detections gets saved with the source_id of the image. Later the cocoapi calculates the map with the gt_boxes from the image with the corresponding source_id.
